### PR TITLE
GAX resource-name changes

### PR DESCRIPTION
**Pull request by chrisdunelm**

### CODE REVIEW

File: src/Google.Api.Gax/IResourceName.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+namespace Google.Api.Gax
+{
+    /// <summary>
+    /// A resource-name.
```
Comment (jskeet) [inline src/Google.Api.Gax/IResourceName.cs]: Any reason for the hyphenation everywhere? I think it's just a resource name...

File: src/Google.Api.Gax/ResourceNameKind.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace Google.Api.Gax
+{
+    /// <summary>
+    /// The kind of resource-name.
+    /// </summary>
+    public enum ResourceNameKind
+    {
+        /// <summary>
+        /// A simple, well-typed, resource name.
```
Comment (jskeet) [inline src/Google.Api.Gax/ResourceNameKind.cs]: Not sure *exactly* what well-typed means here, but I'll let it go :)

File: src/Google.Api.Gax/TemplatedResourceName.cs [hunk 3]
```diff
@@ -107,31 +106,49 @@ public ResourceName(PathTemplate template, params string[] resourceIds)
...
         public override string ToString() => Template.ReplaceParameters(ServiceName, _resourceIds);
+
+        /// <inheritdoc />
+        public override int GetHashCode() => ToString().GetHashCode();
+
+        /// <inheritdoc />
+        public override bool Equals(object obj) => Equals(obj as TemplatedResourceName);
+
+        /// <inheritdoc />
+        public bool Equals(TemplatedResourceName other) => ToString() == other?.ToString();
```
Comment (jskeet) [inline src/Google.Api.Gax/TemplatedResourceName.cs]: Hmm. I wonder whether we should check the template as well. Let's not worry for now.

File: test/Google.Api.Gax.Tests/ResourceNameListTest.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+{
+    public class ResourceNameListTest
+    {
+        private IList<string> List(params string[] items) => new List<string>(items);
+        private ResourceNameList<UnknownResourceName> ResourceList(IList<string> list) =>
+            new ResourceNameList<UnknownResourceName>(list, s => new UnknownResourceName(s));
+
+
+    [Fact]
+    public void Cound()
```
Comment (jskeet) [inline test/Google.Api.Gax.Tests/ResourceNameListTest.cs]: Indentation and Cound => Count.

File: test/Google.Api.Gax.Tests/ResourceNameListTest.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace Google.Api.Gax.Tests
+{
+    public class ResourceNameListTest
+    {
+        private IList<string> List(params string[] items) => new List<string>(items);
```
Comment (jskeet) [inline test/Google.Api.Gax.Tests/ResourceNameListTest.cs]: Any reason not to just use `new[] { "a", "b" }` etc? No need to create a `List<T>`, as `string[]` implements `IList<string>`...

File: test/Google.Api.Gax.Tests/ResourceNameListTest.cs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        private ResourceNameList<UnknownResourceName> ResourceList(IList<string> list) =>
```
Comment (jskeet) [inline test/Google.Api.Gax.Tests/ResourceNameListTest.cs]: Just take params string[] here?

File: test/Google.Api.Gax.Tests/ResourceNameListTest.cs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        [Fact]
+        public void RemoveAt()
+        {
+            var list = List("a", "b");
+            ResourceList(list).RemoveAt(1);
+            Assert.Equal(List("a"), list);
+        }
+
+        [Fact]
+        public void Null()
```
Comment (jskeet) [inline test/Google.Api.Gax.Tests/ResourceNameListTest.cs]: Add_NullElement?

File: test/Google.Api.Gax.Tests/UnknownResourceNameTest.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            Assert.False(name1 != name2);
+            var name3 = UnknownResourceName.Parse("b");
+            Assert.False(name1.GetHashCode() == name3.GetHashCode());
+            Assert.False(name1.Equals(name3));
+            Assert.False(name1.Equals((object)name3));
+            Assert.False(name3.Equals(name1));
+            Assert.False(name3.Equals((object)name1));
+            Assert.False(name1 == name3);
+            Assert.True(name1 != name3);
+
```
Comment (jskeet) [inline test/Google.Api.Gax.Tests/UnknownResourceNameTest.cs]: REmove blank line.

File: src/Google.Api.Gax/IResourceName.cs (hunk 1 above)
Comment (chrisdunelm) [inline src/Google.Api.Gax/IResourceName.cs]: No reason. All removed.

File: src/Google.Api.Gax/ResourceNameKind.cs (hunk 2 above)
Comment (chrisdunelm) [inline src/Google.Api.Gax/ResourceNameKind.cs]: :)

File: src/Google.Api.Gax/TemplatedResourceName.cs (hunk 3 above)
Comment (chrisdunelm) [inline src/Google.Api.Gax/TemplatedResourceName.cs]: Yes, I wondered about this too, but thought it's probably necessary.
We can always revisit if required.

File: test/Google.Api.Gax.Tests/ResourceNameListTest.cs (hunk 5 above)
Comment (chrisdunelm) [inline test/Google.Api.Gax.Tests/ResourceNameListTest.cs]: Some tests need to increase the length of the list.
All tests use this just for consistency.

File: test/Google.Api.Gax.Tests/ResourceNameListTest.cs (hunk 6 above)
Comment (chrisdunelm) [inline test/Google.Api.Gax.Tests/ResourceNameListTest.cs]: Some tests need to have the underlying IList, hence constructing the ResourceNameList in two parts.

File: test/Google.Api.Gax.Tests/ResourceNameListTest.cs (hunk 4 above)
Comment (chrisdunelm) [inline test/Google.Api.Gax.Tests/ResourceNameListTest.cs]: Done.
This was when VS decided it didn't want to help me, and I was having to indent by hand(!)

File: test/Google.Api.Gax.Tests/ResourceNameListTest.cs (hunk 7 above)
Comment (chrisdunelm) [inline test/Google.Api.Gax.Tests/ResourceNameListTest.cs]: Done

File: test/Google.Api.Gax.Tests/UnknownResourceNameTest.cs (hunk 8 above)
Comment (chrisdunelm) [inline test/Google.Api.Gax.Tests/UnknownResourceNameTest.cs]: Done

**Pull request merged**